### PR TITLE
8297729: Replace GrowableArray in ComputeMoveOrder with hash table

### PR DESCRIPTION
**Pull request by JornVernee**
Replaces the GrowableArray 'table' in ComputeMoveOrder with a real hash table.

Through testing, I found that sometimes this array is blown up to several thousand elements, most of which are `NULL`. Using a hash table prevents this large wastage.

I've touched up some of the surrounding code as well. Mostly style changes, but I've also removed the `BasicType` field from the `Move` and `MoveOperation` structs, since it was unused, and added frame data storages to the fast path for stack args as well, since both are allocated on the stack.

Testing: `jdk_foreign` test suite.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297729](https://bugs.openjdk.org/browse/JDK-8297729): Replace GrowableArray in ComputeMoveOrder with hash table

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [40e66a7c](https://git.openjdk.org/jdk/pull/11392/files/40e66a7c29b302b787c0b97bb12c37c07772950e)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11392/head:pull/11392` \
`$ git checkout pull/11392`

Update a local copy of the PR: \
`$ git checkout pull/11392` \
`$ git pull https://git.openjdk.org/jdk pull/11392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11392`

View PR using the GUI difftool: \
`$ git pr show -t 11392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11392.diff">https://git.openjdk.org/jdk/pull/11392.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Seems ok.

File: src/hotspot/share/prims/foreignGlobals.cpp [hunk 1]
```diff
@@ -182,63 +182,75 @@ int JavaCallingConvention::calling_convention(const BasicType* sig_bt, VMStorage
...
+    return static_cast<unsigned int>(vms.type()) ^ vms.index_or_offset();
+  }
+
+  using KillerTable = ResourceHashtable<
+    VMStorage, MoveOperation*,
+    32, // doesn't need to be big. don't have that many argument registers (in known ABIs)
+    AnyObj::RESOURCE_AREA,
+    mtInternal,
+    ComputeMoveOrder::hash,
+    ::operator==
```
[the remark this answers is not in the document]
Comment (JornVernee) [reply src/hotspot/share/prims/foreignGlobals.cpp:200]: This is referencing the global `operator==` defined for `VMStorage` here: https://github.com/openjdk/jdk/blob/6a7a0db587446160e77c9ac225669e5a0c9bc638/src/hotspot/share/prims/vmstorage.hpp#L92

If you like, I can define a wrapper similar to the `hash` function for this as well.

[the remark this answers is not in the document]
Comment (JornVernee) [reply src/hotspot/share/prims/foreignGlobals.cpp:200]: Actually, now that you mention it. I think we need another `equals` function that ignores the `segment_mask_or_size` field just like the `hash` function does.

I will update the PR.

[the remark this answers is not in the document]
Comment (JornVernee) [reply src/hotspot/share/prims/foreignGlobals.cpp]: I've added a specialized `equals` function that ignores the `segment_mask_or_size` field, and changed the template parameter to use that.

**Review by jdksjolen — APPROVED** [review]
Looks good, thank you for the changes.

**Pull request closed without merging**